### PR TITLE
refactor(connlib): polish DNS resource matching

### PR DESCRIPTION
**Pull request by thomaseizinger**
In preparation for implementing #5056, I familiarized myself with the current code and ended up implementing a couple of refactorings.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
Review commit-by-commit is recommended.

File: rust/connlib/tunnel/src/dns.rs [hunk 1]
```diff
@@ -439,41 +421,47 @@ pub fn is_subdomain(name: &DomainName, resource: &str) -> bool {
...
+fn match_domain<T>(name: &DomainName, resources: &HashMap<String, T>) -> Option<T>
+where
+    T: Copy,
+{
+    // Safety: `?` is less than 254 bytes long.
+    const QUESTION_MARK: RelativeName<&'static [u8]> =
+        unsafe { RelativeName::from_octets_unchecked(b"\x01?") };
+    // Safety: `*` is less than 254 bytes long.
+    const WILDCARD: RelativeName<&'static [u8]> =
+        unsafe { RelativeName::from_octets_unchecked(b"\x01*") };
```
Comment (conectado) [inline rust/connlib/tunnel/src/dns.rs:433]: I think we can use [this](https://docs.rs/domain/latest/domain/base/name/struct.RelativeName.html#method.wildcard_ref)

**Review by conectado — APPROVED** [review]
LGTM

File: rust/connlib/tunnel/src/dns.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/dns.rs:433]: It is not a `const fn` unfortunately and I thought that consistency is nicer here.

**Pull request merged**